### ISSUE TITLE
Training SAM3 but freezing specific layers. Want to finetune on new data but keep modifiers

### ISSUE THREAD

**Issue by jlee-detect** [facebookresearch/sam3#455]
I’m looking to fine-tune SAM 3 on a custom dataset of industrial objects that the base model doesn't recognize well (specifically doesn't recognize the word). I have a question about how finetuning affects the model's ability to understand modifiers (adjectives) that aren't in my training data.

My training data is labeled with simple noun phrases (e.g., just forklift or pallet_jack). The labels don't describe the object's state or color.
If I fine-tune on this data, will the model lose its ability to understand adjectives like "yellow" vs "orange" because I'm never reinforcing them during training? Is there a current method to freeze certain parts to maintain SAM3's ability to use adjectives and modifiers like color, shape, size etc.

As a simplified example, if we were using the forklift example, my dataset would have both of these objects labeled as just forklift

<img width="225" height="225" alt="Image" src="https://github.com/user-attachments/assets/51a22d55-f3f4-4cf1-b39b-ba0eabf151c4" />

<img width="275" height="183" alt="Image" src="https://github.com/user-attachments/assets/0972edda-cc2d-4c9a-9ece-3e21da6c1058" />

**Comment by Khadibia** [comment]
I wonder if you moved past this. I am facing the layer freezing challenge over here. In my case, all but the decoder doesn't seem to freeze when i try freezing them. I wonder if you succeeded?

**Comment by 1DoraGON** [comment]
Hi, I used this to freeze what I want in the sam3/train/trainer.py _setup_components()

        self.model = instantiate(self.model_conf, _convert_="all")
        # --- ADD THESE LINES TO FREEZE BACKBONES ---
        # print("Freezing vision and language backbones to save VRAM...")
        # for name, param in self.model.named_parameters():
        #     if "vision_backbone" in name:
        #     # if "vision_backbone" in name or "language_backbone" in name:
        #         param.requires_grad = False
        # # -------------------------------------------

        # NEW — unfreeze last 4 ViT blocks (28-31), keep rest frozen
        # print("Partially freezing backbones...")
        # UNFREEZE_BLOCKS = {"28", "29", "30", "31"}
        # for name, param in self.model.named_parameters():
        #     if "language_backbone" in name:
        #         param.requires_grad = False  # keep language fully frozen, small dataset
        #     elif "vision_backbone" in name:
        #     # if "vision_backbone" in name:
        #         # check if this param belongs to one of the last 4 blocks
        #         is_unfrozen = any(
        #             f"trunk.blocks.{b}." in name for b in UNFREEZE_BLOCKS
        #         )
        #         param.requires_grad = is_unfrozen
        #     # everything else (decoder/transformer) stays trainable by default

        # Partially unfreeze backbones — tune block ranges here
        print("Partially freezing backbones...")

        UNFREEZE_VIS_BLOCKS = {"24", "25", "26", "27", "28", "29", "30", "31"}  # last 8 of 32
        UNFREEZE_LANG_BLOCKS = {"22", "23"}  # last 2 of 24

        for name, param in self.model.named_parameters():
            if "language_backbone" in name:
                is_unfrozen = any(
                    f"resblocks.{b}." in name for b in UNFREEZE_LANG_BLOCKS
                )
                param.requires_grad = is_unfrozen

            elif "vision_backbone" in name:
                is_unfrozen = any(
                    f"trunk.blocks.{b}." in name for b in UNFREEZE_VIS_BLOCKS
                )
                param.requires_grad = is_unfrozen

            # everything else (decoder/transformer) stays trainable by default

        # Print summary
        total = sum(p.numel() for p in self.model.parameters())
        trainable = sum(p.numel() for p in self.model.parameters() if p.requires_grad)
        print(f"Trainable: {trainable/1e6:.1f}M / {total/1e6:.1f}M ({100*trainable/total:.1f}%)")
        print_model_summary(self.model)

        self.loss = None
